qiskit/quantum_info/operators/op_shape.py

```diff
@@ -0,0 +1,496 @@
+        ndim = len(shape)
+        if ndim > 2:
+            if raise_exception:
+                raise QiskitError(
+                    "Input shape is not 1 or 2-dimensional (shape = {})".format(shape))
+            return False
+
+        if self._dims_l:
+            if reduce(mul, self._dims_l) != shape[0]:
+                if raise_exception:
+                    raise QiskitError(
+                        "Output dimensions do not match matrix shape "
+                        "({} != {})".format(reduce(mul, self._dims_l), shape[0]))
+                return False
+        elif shape[0] != 2 ** self._num_qargs_l:
+            if raise_exception:
+                raise QiskitError("Number of left qubits does not match matrix shape")
+            return False
+
+        if ndim == 2:
+            if self._dims_r:
+                if reduce(mul, self._dims_r) != shape[1]:
+                    if raise_exception:
+                        raise QiskitError(
+                            "Input dimensions do not match matrix shape "
+                            "({} != {})".format(reduce(mul, self._dims_r), shape[1]))
+                    return False
+            elif shape[1] != 2 ** self._num_qargs_r:
+                if raise_exception:
+                    raise QiskitError("Number of right qubits does not match matrix shape")
+                return False
+        elif self._dims_r or self._num_qargs_r:
+            if raise_exception:
+                raise QiskitError('Input dimension should be empty for vector shape.')
+            return False
+
+        return True
+
+    @classmethod
+    def auto(cls, shape=None, dims_l=None, dims_r=None, dims=None,
+             num_qubits_l=None, num_qubits_r=None, num_qubits=None):
+        """Construct TensorShape with automatic checking of qubit dimensions"""
+        if dims and (dims_l or dims_r):
+            raise QiskitError(
+                "`dims` kwarg cannot be used with `dims_l` or `dims_r`")
+        if num_qubits and (num_qubits_l or num_qubits_r):
+            raise QiskitError(
+                "`num_qubits` kwarg cannot be used with `num_qubits_l` or `num_qubits_r`")
+
+        if num_qubits:
+            num_qubits_l = num_qubits
+            num_qubits_r = num_qubits
+        if dims:
+            dims_l = dims
+            dims_r = dims
+
+        if num_qubits_r and num_qubits_l:
+            matrix_shape = cls(num_qargs_l=num_qubits_r,
+                               num_qargs_r=num_qubits_l)
+        else:
+            ndim = len(shape) if shape else 0
+            if dims_r is None and num_qubits_r is None and ndim > 1:
+                dims_r = shape[1]
+
+            if dims_l is None and num_qubits_l is None and ndim > 0:
+                dims_l = shape[0]
+
+            if num_qubits_r is None:
+                if isinstance(dims_r, Integral):
+                    nq = int(log2(dims_r))
+                    if dims_r == 2 ** nq:
```

Comment:
Is there a better way to verify that `dim_r` is a power of 2?
See: https://stackoverflow.com/questions/57025836/how-to-check-if-a-given-number-is-a-power-of-two

Reply:
Interesting! I hadn't thought too hard about it, but something that works directly with ints would be much better